spec/unit/graph/simple_graph_spec.rb

```diff
@@ -681,16 +686,20 @@ def graph_to_yaml(graph, which_format)
       end
 
       it "should be able to serialize a graph where the vertices contain backreferences to the graph (#{which_format} format)" do
+        pending('Make SimpleGraph a known to the Puppet Type System')
```

Comment:
None of those have tests where they are made to contain an object that in turn points back to themselves with the requirement that this self-reference is resolved during deserialization. In order to do that with Ruby < 2.3, we will either need to go back to the `#to_yaml_properties` approach, or use an alternative. The `#encode_with` seems to work well with self-references as long as they don't reference the top object. When that happens Rubies < 2.3 fails.

That leaves us with three alternatives:
1. Keep ` #to_yaml_properties` for the sole sake of being able to do this.
2. Use an alternative approach. Registering `SimpleGraph` in the Puppet Type system and use rich_data solves this.
3. Conclude that this self-reference capability is not needed and remove the test. That would be my take.

The Catalog rewrites the `#to_data_hash` completely and does not use the functionality in `SimpleGraph` at all. So who benefits from this self-referencing capability anyway?